src/pybind/mgr/restful/module.py

```diff
@@ -0,0 +1,487 @@
+            "desc": "Delete an API key with this name",
+            "perm": "rw"
+        },
+        {
+            "cmd": "list_keys",
+            "desc": "List all API keys",
+            "perm": "rw"
+        },
+    ]
+
+    def __init__(self, *args, **kwargs):
+        super(Module, self).__init__(*args, **kwargs)
+        global instance
+        instance = self
+
+        self.requests = []
+        self.requests_lock = threading.RLock()
+
+        self.keys = {}
+        self.disable_auth = False
+
+        self.server = None
+
+
+    def serve(self):
+        try:
+            self._serve()
+        except:
+            self.log.error(str(traceback.format_exc()))
+
+
+    def _serve(self):
+        # Load stored authentication keys
+        self.keys = self.get_config_json("keys") or {}
+
+        jsonify._instance = jsonify.GenericJSON(
+            sort_keys=True,
+            indent=4,
+            separators=(',', ': '),
+        )
+
+        cert = self.get_config_json("cert") or '/etc/ceph/ceph-mgr-restful.crt'
+        pkey = self.get_config_json("pkey") or '/etc/ceph/ceph-mgr-restful.key'
+
+        # Create the HTTPS werkzeug server serving pecan app
+        self.server = make_server(
+            host='0.0.0.0',
+            port=8003,
+            app=make_app(
+                root='restful.api.Root',
+                hooks = lambda: [ErrorHook()],
+            ),
+            ssl_context=(cert, pkey),
+        )
+
+        self.server.serve_forever()
+
+
+    def shutdown(self):
+        try:
+            if self.server:
+                self.server.shutdown()
+        except:
+            self.log.error(str(traceback.format_exc()))
+
+
+    def notify(self, notify_type, tag):
+        try:
+            self._notify(notify_type, tag)
+        except:
+            self.log.error(str(traceback.format_exc()))
+
+
+    def _notify(self, notify_type, tag):
+        if notify_type == "command":
+            # we can safely skip all the sequential commands
+            if tag == 'seq':
+                return
+
+            request = filter(
+                lambda x: x.is_running(tag),
+                self.requests)
+
+            if len(request) != 1:
+                self.log.warn("Unknown request '%s'" % str(tag))
+                return
+
+            request = request[0]
+            request.finish(tag)
+            if request.is_ready():
+                request.next()
+        else:
+            self.log.debug("Unhandled notification type '%s'" % notify_type)
+
+
+    def handle_command(self, command):
+        self.log.warn("Handling command: '%s'" % str(command))
+        if command['prefix'] == "create_key":
+            if command['key_name'] in self.keys:
+                return 0, self.keys[command['key_name']], ""
+
+            else:
+                self.keys[command['key_name']] = str(uuid4())
+                self.set_config_json('keys', self.keys)
```

Comment:
hi, a quick question: `self.set_config_json` probably run failed, like bad authorization. should we process this case here?